### PR TITLE
feat: Added support of Poly Loss

### PR DESCRIPTION
**Pull request by frgfm**
As discussed in #6439, this PR adds support for the poly loss. A few things are left to decide / do:
- [ ] decide whether a module API should also be added (easy to add in this PR)
- [ ] specify the depths of the unittests (I've checked the one of the focal loss, it's quite extensive. how much should we do? checking trivial vaues?)

cc @datumbox

### CODE REVIEW

File: torchvision/ops/poly_loss.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+from ..utils import _log_api_usage_once
+
+
+def poly_loss(
+    x: Tensor,
+    target: Tensor,
+    eps: float = 2.0,
+    weight: Optional[Tensor] = None,
+    ignore_index: int = -100,
+    reduction: str = "mean",
```
Comment (oke-aditya) [inline torchvision/ops/poly_loss.py:15]: I also had a thought of making this into an Enum say LossReductionMode.

File: torchvision/ops/poly_loss.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+from ..utils import _log_api_usage_once
+
+
+def poly_loss(
+    x: Tensor,
+    target: Tensor,
+    eps: float = 2.0,
+    weight: Optional[Tensor] = None,
+    ignore_index: int = -100,
```
Comment (oke-aditya) [inline torchvision/ops/poly_loss.py:14]: this is just to have default ignore as false right? Why not set it to None and check if its not None?

File: torchvision/ops/poly_loss.py (hunk 1 above)
Comment (datumbox) [reply torchvision/ops/poly_loss.py:15]: Though I do prefer enums to strings, keeping this a string will ensure strong alignment with Core and will assist on a possible upstreaming on the future.

File: torchvision/ops/poly_loss.py (hunk 2 above)
Comment (datumbox) [reply torchvision/ops/poly_loss.py:14]: My understanding is he is following very closely the default values of established losses in Core ([example](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html)). I think that's a good idea.

**Review by oke-aditya — COMMENTED** [review]
Sorry for early poke but a few thoughts @frgfm

File: torchvision/ops/poly_loss.py (hunk 2 above)
Comment (oke-aditya) [reply torchvision/ops/poly_loss.py:14]: aah I didn't know about this. Nice

File: torchvision/ops/poly_loss.py (hunk 1 above)
Comment (oke-aditya) [reply torchvision/ops/poly_loss.py:15]: I'm suggesting to support both strings and LossReductionMode. But that's ok,

Comment (frgfm) [reply torchvision/ops/poly_loss.py:15]: I hesitated as well, but went for the same option as core

File: torchvision/ops/poly_loss.py (hunk 2 above)
Comment (frgfm) [reply torchvision/ops/poly_loss.py:14]: Yup that's what I was going for

File: torchvision/ops/poly_loss.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        if weight.type() != x.data.type():
+            weight = weight.type_as(x.data)
+        logpt = weight.gather(0, target.data.view(-1)) * logpt
+
+    # Loss reduction
+    if reduction == "sum":
+        loss = loss[valid_idxs].sum()
+    elif reduction == "mean":
+        loss = loss[valid_idxs].mean()
+    else:
```
[the remark this answers is not in the document]
Comment (frgfm) [reply torchvision/ops/poly_loss.py:62]: I think a `ValueError` in the constructor is much better indeed

[the remark this answers is not in the document]
Comment (oke-aditya) [reply torchvision/ops/poly_loss.py:62]: We don't have any class based module. Only functional ones for losses. 

cc @datumbox for this this point. I think we can throw value error here too.

[the remark this answers is not in the document]
Comment (datumbox) [reply torchvision/ops/poly_loss.py:62]: Just being mindful that @frgfm has marked the PR as draft, so I don't want to snoop around too much before he is ready. 

But I think it would be OK to throw a value exception here as this is aligned with a [practice](https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/nn/_reduction.py#L7-L20) from PyTorch core. We possibly try to go for similar look-and-feel so that if we end up moving it on Core, the implementation behaves similarly.

**Comment by datumbox** [comment]
@frgfm Just wanted to check in and ensure you are not blocked or need any assistance. Let me know if you do. :)

File: torchvision/ops/poly_loss.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (frgfm) [reply torchvision/ops/poly_loss.py]: Just added a ValueError!

**Comment by frgfm** [comment]
Hello there :) 

My bad, I just caught up with the PR! I left it as a draft because the missing part is the unit test: I saw that the ones for the focal loss are quite extensive. Do I need to be that thorough?

**Comment by datumbox** [comment]
@frgfm Ah yes the unit-tests. I would recommend following the pattern that we have on existing tests and have a similar coverage. Hopefully following the existing infra will make your work easier. The reason we really need them is because previously we spotted a few nasty bugs on some of the losses. We are only couple of weeks prior cutting the branch for the upcoming release so if we want to include this feature we need to have elevated degree of confidence.
